File: lib/src/main/java/com/walmartlabs/components/scheduler/core/BucketManager.java
package com.walmartlabs.components.scheduler.core;

import com.google.common.base.Function;
import com.google.common.collect.HashMultimap;
import com.google.common.collect.Multimap;
import com.google.common.util.concurrent.FutureCallback;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.common.util.concurrent.ListenableScheduledFuture;
import com.hazelcast.core.IMap;
import com.walmart.gmp.ingestion.platform.framework.shutdown.SupportsShutdown;
import com.walmart.gmp.ingestion.platform.framework.data.core.DataManager;
import com.walmart.gmp.ingestion.platform.framework.data.core.Selector;
import com.walmart.marketplace.messages.v1_bigben.EventResponse.Status;
import com.walmartlabs.components.scheduler.entities.Bucket;
import com.walmartlabs.components.scheduler.entities.BucketDO;
import com.walmartlabs.components.scheduler.entities.Event;
import com.walmartlabs.components.scheduler.entities.EventDO.EventKey;
import com.walmartlabs.components.scheduler.entities.EventLookup;
import org.apache.commons.lang3.tuple.Pair;
import org.apache.log4j.Logger;

import java.time.ZonedDateTime;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

import static com.google.common.collect.Lists.newArrayList;
import static com.google.common.util.concurrent.Futures.*;
import static com.walmart.gmp.ingestion.platform.framework.core.Props.PROPS;
import static com.walmart.gmp.ingestion.platform.framework.shutdown.ShutdownRegistry.SHUTDOWN_REGISTRY;
import static com.walmart.gmp.ingestion.platform.framework.data.core.Selector.fullSelector;
import static com.walmart.marketplace.messages.v1_bigben.EventResponse.Status.*;
import static com.walmart.platform.soa.common.exception.util.ExceptionUtil.getRootCause;
import static com.walmartlabs.components.scheduler.core.ScheduleScanner.SCHEDULER;
import static com.walmartlabs.components.scheduler.utils.TimeUtils.nowUTC;
import static java.lang.Runtime.getRuntime;
import static java.lang.String.format;
import static java.util.concurrent.TimeUnit.SECONDS;
import static java.util.stream.Collectors.toList;

/**
 * Created by smalik3 on 6/29/16
 */
public class BucketManager implements SupportsShutdown {

    static final Logger L = Logger.getLogger(BucketManager.class);

    private final int maxBuckets;
    private final int maxProcessingTime;
    private final DataManager<ZonedDateTime, Bucket> dataManager;
    private final DataManager<EventKey, Event> eventDataManager;
    private final int lookbackRange;
    private final Map<ZonedDateTime, BucketSnapshot> buckets = new ConcurrentHashMap<>();
    private final int bucketWidth;
    private volatile BucketsLoader bucketsLoader;
    private final StatusSyncer statusSyncer;
    private final CheckpointHelper checkpointHelper;
    private final IMap<ZonedDateTime, Bucket> map;
    private final int shardSize;

    @SuppressWarnings("unchecked")
    public BucketManager(int maxBuckets, int maxProcessingTime, DataManager<?, ?> dm, int bucketWidth,
                         int checkpointInterval, TimeUnit checkpointUnit, int lookbackRange, IMap<ZonedDateTime, Bucket> map) {
        this.maxBuckets = maxBuckets;
        this.maxProcessingTime = maxProcessingTime;
        this.dataManager = (DataManager<ZonedDateTime, Bucket>) dm;
        this.eventDataManager = (DataManager<EventKey, Event>) dm;
        this.lookbackRange = lookbackRange;
        this.bucketWidth = bucketWidth;
        this.shardSize = PROPS.getInteger("event.shard.size", 1000);
        this.checkpointHelper = new CheckpointHelper(this);
        this.statusSyncer = new StatusSyncer(dataManager, eventDataManager);
        this.map = map;
        L.info(format("saving checkpoint every %d %s", checkpointInterval, checkpointUnit));
        SCHEDULER.scheduleAtFixedRate(this::saveCheckpoint, checkpointInterval, checkpointInterval, checkpointUnit);
        try {
            L.info("loading the previously saved checkpoint, if any");
            transform(checkpointHelper.loadCheckpoint(), (Function<Map<ZonedDateTime, BucketSnapshot>, Object>) data -> {
                L.info("loaded previous checkpoint: " + data);
                buckets.putAll(data);
                purgeIfNeeded();
                return null;
            }).get();
        } catch (Exception e) {
            L.error("could not load previous checkpoint", e);
        }
        getRuntime().addShutdownHook(new Thread(() -> {
            L.info("saving checkpoint during shutdown");
            saveCheckpoint();
        }));
        SHUTDOWN_REGISTRY.register(this);
    }

    private static final Selector<ZonedDateTime, Bucket> selector = fullSelector(nowUTC());

    ListenableFuture<Multimap<ZonedDateTime, Integer>> getProcessableShardsForOrBefore(ZonedDateTime bucketId) {
        if (bucketsLoader == null) {
            L.info("starting the background load of previous buckets");
            final Integer fetchSize = PROPS.getInteger("buckets.background.load.fetch.size", 10);
            bucketsLoader = new BucketsLoader(lookbackRange, fetchSize,
                    b -> buckets.put(b.id(), new BucketSnapshot(b.id(), b.getCount(), shardSize, b.getStatus())),
                    buckets::containsKey, bucketWidth, dataManager, bucketId);
            bucketsLoader.start();
        }

        final ListenableFuture<Bucket> f = dataManager.getAsync(bucketId, selector);//adapt(map.getAsync(bucketId));
        final Multimap<ZonedDateTime, Integer> processableShards = HashMultimap.create();
        addCallback(f, new FutureCallback<Bucket>() {
            @Override
            public void onSuccess(Bucket bucket) {
                bucket = bucket == null ? createEmptyBucket(bucketId) : bucket;
                if (buckets.putIfAbsent(bucketId, new BucketSnapshot(bucketId, bucket.getCount(), shardSize, bucket.getStatus())) != null) {
                    L.warn(format("bucket with id %s already existed in the cache, this is highly unusual", bucketId));
                }
                buckets.entrySet().stream().filter(e -> e.getValue().awaiting().cardinality() > 0).forEach(e ->
                        e.getValue().awaiting().stream().forEach(s -> processableShards.put(e.getKey(), s)));
                L.info(format("processable shards at bucket: %s, are => %s", bucketId, processableShards));
                if (!processableShards.containsKey(bucketId)) {
                    L.info(format("no events in the bucket: %s", bucketId));
                }
            }

            @Override
            public void onFailure(Throwable t) {
                L.info(format("error in loading bucket: %s, will be retried again during next scan", bucketId), t);
            }
        });
        return transform(f, (Function<Bucket, Multimap<ZonedDateTime, Integer>>) $ -> processableShards);
    }

    synchronized void registerForProcessing(Collection<Pair<ZonedDateTime, Integer>> pairs) {
        pairs.forEach(p -> buckets.get(p.getLeft()).processing(p.getRight()));
        purgeIfNeeded();
        startShardsTimer(pairs);
    }

    private ListenableScheduledFuture<?> startShardsTimer(Collection<Pair<ZonedDateTime, Integer>> pairs) {
        final List<String> shards = pairs.stream().sorted().map(p -> p.getLeft() + "[" + p.getRight() + "]").collect(toList());
        L.info(format("starting processing timer for shards: %s", shards));
        return SCHEDULER.schedule(() -> checkShardsStatus(pairs, shards), maxProcessingTime, SECONDS);
    }

    private synchronized void checkShardsStatus(Collection<Pair<ZonedDateTime, Integer>> pairs, List<String> shards) {
        try {
            for (Pair<ZonedDateTime, Integer> pair : pairs) {
                final ZonedDateTime bucketId = pair.getLeft();
                final int shard = pair.getRight();
                final BucketSnapshot bd = buckets.get(bucketId);
                if (bd != null && bd.processing().get(shard)) {
                    L.warn(format("bulk timer for shard: %s[%d] expired, marking the shard as failure", bucketId, shard));
                    bd.done(shard, ERROR);
                }
            }
        } catch (Exception e) {
            L.error(format("error in timing out the shards for processing, shards: %s", shards), e);
        }
    }

    synchronized void shardDone(ZonedDateTime bucketId, Integer shard, Status status) {
        final BucketSnapshot bd = buckets.get(bucketId);
        if (bd == null) {
            L.warn(format("bucket %s not found in cache, might have been purged, ignoring this call", bucketId));
            return;
        }
        bd.done(shard, status);
        /*if (bd.processing().cardinality() == 0) {
            final Status bucketStatus = bd.awaiting().cardinality() == 0 ? PROCESSED : ERROR;
            L.info(format("all shards done for bucket: %s, marking it %s", bucketId, bucketStatus));
            bucketProcessed(bucketId, bucketStatus);
        }*/
    }

    private static final ListenableFuture<Bucket> NO_OP = immediateFuture(null);

    synchronized ListenableFuture<Bucket> bucketProcessed(ZonedDateTime bucketId, Status status) {
        final BucketSnapshot bd = buckets.get(bucketId);
        if (bd == null) {
            L.warn(format("bucket %s not found in cache, this is extremely unusual", bucketId));
            return NO_OP;
        }
        bd.processing().clear();
        if (status == PROCESSED) {
            L.info(format("bucket %s done, marking it as %s, all shards done", bucketId, status));
            bd.awaiting().clear();
        } else if (status == ERROR)
            L.warn(format("bucket %s done, marking it as %s, failed shards are: %s", bucketId, status, bd.awaiting()));
        return statusSyncer.syncBucket(bucketId, status, true);
    }

    private synchronized void purgeIfNeeded() {
        try {
            if (buckets.size() <= maxBuckets) {
                L.info("nothing to purge");
                return;
            }
            L.debug("initiating purge check for buckets: " + this.buckets);
            addCallback(successfulAsList(newArrayList(this.buckets.keySet()).stream().
                    sorted().limit(buckets.size() - maxBuckets).map(b -> {
                //buckets.get(b).count() > 0 || buckets.get(b).processing().cardinality() > 0
                final BucketSnapshot bs = this.buckets.get(b);
                if (bs.processing().cardinality() > 0) {
                    L.info(format("skipping purge of bucket %s, shards are still being processed", b));
                    return NO_OP;
                } else {
                    L.debug("purging bucket snapshot: " + bs);
                    buckets.remove(b);
                    if (bs.count() == 0) {
                        return immediateFuture(createEmptyBucket(b));
                    } else {
                        if (bs.awaiting().cardinality() == 0) {
                            return statusSyncer.syncBucket(b, PROCESSED, false);
                        } else {
                            return statusSyncer.syncBucket(b, ERROR, false);
                        }
                    }
                }
            }).collect(toList())), new FutureCallback<List<Bucket>>() {
                @Override
                public void onSuccess(List<Bucket> l) {
                    L.info("purged buckets: " + l.stream().filter(b -> b != null).collect(toList()));
                }

                @Override
                public void onFailure(Throwable t) {
                    L.error("error in purging buckets", getRootCause(t));
                }
            });
        } catch (Exception e) {
            L.error("error in purging snapshots", getRootCause(e));
        }
    }

    synchronized ListenableFuture<EventLookup> saveCheckpoint() {
        purgeIfNeeded();
        return checkpointHelper.saveCheckpoint(buckets);
    }

    int getMaxProcessingTime() {
        return maxProcessingTime;
    }

    DataManager<EventKey, Event> getEventDataManager() {
        return eventDataManager;
    }

    StatusSyncer getStatusSyncer() {
        return statusSyncer;
    }

    static Bucket createEmptyBucket(ZonedDateTime bId) {
        final BucketDO bucketDO = new BucketDO();
        bucketDO.setId(bId);
        bucketDO.setStatus(EMPTY.name());
        bucketDO.setCount(0);
        return bucketDO;
    }

    @Override
    public int priority() {
        return 1;
    }

    @Override
    public String name() {
        return "BucketManager";
    }

    @Override
    public ListenableFuture<?> shutdown() {
        return saveCheckpoint();
    }
}

File: lib/src/main/kotlin/com/walmartlabs/opensource/bigben/core/BucketManager.kt
package com.walmartlabs.opensource.bigben.core

import com.google.common.base.Throwables.getRootCause
import com.google.common.collect.HashMultimap
import com.google.common.collect.Multimap
import com.google.common.util.concurrent.Futures.immediateFuture
import com.google.common.util.concurrent.ListenableFuture
import com.google.common.util.concurrent.ListenableScheduledFuture
import com.google.common.util.concurrent.MoreExecutors.listeningDecorator
import com.walmartlabs.opensource.bigben.entities.Bucket
import com.walmartlabs.opensource.bigben.entities.EventStatus
import com.walmartlabs.opensource.bigben.entities.EventStatus.*
import com.walmartlabs.opensource.bigben.extns.domainProvider
import com.walmartlabs.opensource.bigben.extns.fetch
import com.walmartlabs.opensource.core.*
import com.walmartlabs.opensource.core.utils.Props
import java.lang.Runtime.getRuntime
import java.time.ZonedDateTime
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.ThreadFactory
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeUnit.SECONDS
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Predicate
import kotlin.Comparator

/**
 * Created by smalik3 on 2/21/18
 */
class BucketManager(private val maxBuckets: Int, private val maxProcessingTime: Int, private val bucketWidth: Int,
                    checkpointInterval: Long, checkpointUnit: TimeUnit) {

    companion object {
        private val l = logger<BucketManager>()
        private val index = AtomicInteger()
        val scheduler = listeningDecorator(ScheduledThreadPoolExecutor(4, ThreadFactory { Thread(it, "BucketManager-${index.getAndIncrement()}") }))!!

        internal fun emptyBucket(bucketId: ZonedDateTime) = domainProvider<Bucket>().let { it.raw(it.selector(Bucket::class.java)).apply { this.bucketId = bucketId; count = 0L; status = EMPTY } }
    }

    private val shardSize = Props.int("event.shard.size", 1000)
    private val checkpointHelper = CheckpointHelper()
    private val statusSyncer = StatusSyncer()
    private val buckets = ConcurrentHashMap<ZonedDateTime, BucketSnapshot>()

    @Volatile private var bucketsLoader: BucketsLoader? = null

    init {
        if (l.isDebugEnabled) l.debug("saving checkpoint every {} {}", checkpointInterval, checkpointUnit)
        scheduler.scheduleAtFixedRate({ saveCheckpoint() }, checkpointInterval, checkpointInterval, checkpointUnit)
        if (l.isDebugEnabled) l.debug("loading the previously saved checkpoint, if any")
        try {
            checkpointHelper.loadCheckpoint().transform {
                if (l.isDebugEnabled) l.debug("loaded previous checkpoint: {}")
                buckets.putAll(it!!)
                purgeIfNeeded()
            }.get()
        } catch (e: Exception) {
            l.error("could not load previous checkpoint", getRootCause(e))
        }
        getRuntime().addShutdownHook(Thread {
            if (l.isInfoEnabled) l.info("saving checkpoint during shutdown")
            saveCheckpoint()
        })
    }

    fun getProcessableShardsForOrBefore(bucketId: ZonedDateTime): ListenableFuture<out Multimap<ZonedDateTime, Int>> {
        if (bucketsLoader == null) {
            if (l.isInfoEnabled) l.info("starting the background load of previous buckets")
            val fetchSize = Props.int("buckets.background.load.fetch.size", 10)
            bucketsLoader = BucketsLoader(maxBuckets - 1, fetchSize, Predicate { buckets.containsKey(it) }, bucketWidth, bucketId) {
                buckets.put(it.bucketId!!, BucketSnapshot.with(it.bucketId!!, it.count!!, shardSize, it.status!!))
            }.apply { run() }
        }
        return HashMultimap.create<ZonedDateTime, Int>().let { shards ->
            fetch<Bucket> { it.bucketId = bucketId }.transform {
                val bucket = it ?: emptyBucket(bucketId)
                if (buckets.putIfAbsent(bucketId, BucketSnapshot.with(bucketId, bucket.count!!, shardSize, bucket.status!!)) != null) {
                    l.warn("bucket with bucketId {} already existed in the cache, this is highly unusual", bucketId)
                }
                buckets.entries.filter { e -> e.value.awaiting.cardinality() > 0 }.forEach { e -> e.value.awaiting.stream().forEach { s -> shards.put(e.key, s) } }
                if (l.isDebugEnabled) l.debug("processable shards at bucket: {}, are => {}", bucketId, shards)
                if (!shards.containsKey(bucketId)) {
                    if (l.isDebugEnabled) l.debug("no events in the bucket: {}", bucketId)
                }
                shards
            }.catching { e -> shards.also { l.warn("error in loading bucket: {}, will be retried again during next scan", bucketId, e.rootCause()) } }
        }
    }

    internal fun registerForProcessing(pairs: Collection<Pair<ZonedDateTime, Int>>) {
        pairs.forEach { p -> buckets[p.first]!!.processing(p.second) }
        purgeIfNeeded()
        startShardsTimer(pairs)
    }

    private fun startShardsTimer(pairs: Collection<Pair<ZonedDateTime, Int>>): ListenableScheduledFuture<*> {
        return pairs.sortedWith(Comparator { p1, p2 ->
            p1.first.compareTo(p2.first).
                    let { if (it != 0) it else p1.second.compareTo(p2.second) }
        }).map { "${it.first}[${it.second}]" }.toList().let {
            if (l.isDebugEnabled) l.debug("starting processing timer for shards: {}", it)
            scheduler.schedule({ checkShardsStatus(pairs, it) }, maxProcessingTime.toLong(), SECONDS)
        }
    }

    @Synchronized private fun checkShardsStatus(pairs: Collection<Pair<ZonedDateTime, Int>>, shards: List<String>) {
        try {
            pairs.forEach {
                val bd = buckets[it.first]
                if (bd != null && bd.processing.get(it.second)) {
                    l.warn("bulk timer for shard: {}[{}] expired, marking the shard as failure", it.first, it.second)
                    bd.done(it.second, ERROR)
                }
            }
        } catch (e: Exception) {
            l.error("error in timing out the shards for processing, shards: {}", shards, e)
        }
    }

    @Synchronized internal fun shardDone(bucketId: ZonedDateTime, shard: Int?, status: EventStatus) {
        val bd = buckets[bucketId]
        if (bd == null) {
            l.warn("bucket {} not found in cache, might have been purged, ignoring this call", bucketId)
            return
        }
        bd.done(shard!!, status)
    }

    private val noOp = immediateFuture<Bucket>(null)

    @Synchronized internal fun bucketProcessed(bucketId: ZonedDateTime, status: EventStatus): ListenableFuture<Bucket> {
        val bd = buckets[bucketId]
        if (bd == null) {
            l.warn("bucket {} not found in cache, this is extremely unusual", bucketId)
            return noOp
        }
        bd.processing.clear()
        if (status == PROCESSED) {
            if (l.isInfoEnabled) l.info("bucket {} done, marking it as {}, all shards done", bucketId, status)
            bd.awaiting.clear()
        } else if (status == ERROR)
            l.warn("bucket {} done, marking it as {}, failed shards are: {}", bucketId, status, bd.awaiting)
        return statusSyncer.syncBucket(bucketId, status, true)
    }

    @Synchronized private fun saveCheckpoint() {
        try {
            purgeIfNeeded()
            checkpointHelper.saveCheckpoint(buckets)
        } catch (e: Throwable) {
            l.error("failed to save checkpoint", e)
        }
    }

    fun purgeIfNeeded() {
        when {
            buckets.size <= maxBuckets -> if (l.isDebugEnabled) l.debug("nothing to purge")
            else -> {
                if (l.isDebugEnabled) l.debug("initiating purge check for buckets: {}", this.buckets)
                buckets.keys.sorted().take(buckets.size - maxBuckets).map { b ->
                    buckets[b]!!.let {
                        if (it.processing.cardinality() > 0) {
                            if (l.isDebugEnabled) l.debug("skipping purge of bucket {}, shards are still being processed", b)
                            immediateFuture(it)
                        } else {
                            if (l.isDebugEnabled) l.debug("purging bucket snapshot: {}", it)
                            val bs = buckets.remove(it.id)
                            when {
                                it.count == 0L -> immediateFuture(it)
                                it.awaiting.cardinality() == 0 -> {
                                    if (l.isDebugEnabled) l.debug("bucket {} is processed", b)
                                    statusSyncer.syncBucket(b, PROCESSED, false).transform { bs }
                                }
                                else -> {
                                    l.warn("bucket {} is marked error as final status", b)
                                    statusSyncer.syncBucket(b, ERROR, false).transform { bs }
                                }
                            }
                        }
                    }
                }.reduce().done({ l.error("error in purging snapshots", getRootCause(it!!)) }) {
                    if (l.isInfoEnabled) l.info("purged buckets: {}", it?.map { it!!.id })
                }
            }
        }
    }

}